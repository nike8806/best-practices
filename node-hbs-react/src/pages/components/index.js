export { default as WorkEmail } from './WorkEmail/WorkEmail.js';
