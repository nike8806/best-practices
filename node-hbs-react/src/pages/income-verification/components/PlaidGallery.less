.plaid-gallery__bank-logo:hover,
.plaid-gallery__find-bank:hover,
.plaid-gallery__find-bank a:hover {
  background-color: #f9f9f9;
  color: #707884;
  cursor: pointer;
  text-decoration: none;
}

.plaid-gallery__bank-logo {
  background-color: #ffffff;
  border: 1px solid #e2e2e2;
  border-radius: 5px;
  display: block;
  margin: 2px;
  background-position: center;
  background-repeat: no-repeat;
  background-size: 65%;
  height: 115px;
}

.plaid-gallery__find-bank {
  background-color: #ffffff;
  border: 1px solid #e2e2e2;
  border-radius: 5px;
}

.plaid-gallery__bank-logo--chase {
  background-image: url('../images/chase-logo.svg');
}

.plaid-gallery__bank-logo--wells {
  background-image: url('../images/wells-fargo-logo.svg');
}

.plaid-gallery__bank-logo--bofa {
  background-image: url('../images/bofa-logo.svg');
}

.plaid-gallery__bank-logo--pnc {
  background-image: url('../images/pnc-logo.svg');
}

.plaid-gallery__bank-logo--us {
  background-image: url('../images/us-bank-logo.svg');
}

.plaid-gallery__bank-logo--usaa {
  background-image: url('../images/usaa-logo.svg');
  background-size: 50%;
}

@media (min-width: 376px) and (max-width:767px) {
  .plaid-gallery__bank-logo {
      background-size: 40%;
      height: 120px;
  }

  .plaid-gallery__bank-logo--usaa {
      background-size: 26%;
  }
}

@media (min-width: 768px) {
  .plaid-gallery__bank-logo {
      background-size: 55%;
  }

  .plaid-gallery__bank-logo--usaa {
      background-size: 40%;
  }
}
