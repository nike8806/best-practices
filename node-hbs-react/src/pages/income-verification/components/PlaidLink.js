import React, { Component } from 'react';
import PropTypes from 'prop-types';

class PlaidLink extends Component {
  static defaultProps = {
    institution: null,
    onBeforeOpen: null,
    plaidInstance: null,
    className: null
  };

  static propTypes = {
    institution: PropTypes.string,
    onBeforeOpen: PropTypes.func,
    plaidInstance: PropTypes.shape({
      open: PropTypes.func.isRequired
    }),
    children: PropTypes.node,
    className: PropTypes.string
  }

  handleClick = () => {
    const {
      onBeforeOpen,
      institution,
      plaidInstance
    } = this.props;

    if (onBeforeOpen) {
      onBeforeOpen({institution});
    }

    if (plaidInstance) {
      plaidInstance.open(institution);
    }
  }

  render() {
    const {
      children,
      onBeforeOpen, // eslint-disable-line no-unused-vars
      plaidInstance, // eslint-disable-line no-unused-vars
      institution, // eslint-disable-line no-unused-vars
      className,
      ...linkAttributes
    } = this.props;

    return (
      <a
        {...linkAttributes}
        onClick={ this.handleClick }
        className={ className }
      >
        { children }
      </a>
    );
  }
}

export default PlaidLink;
