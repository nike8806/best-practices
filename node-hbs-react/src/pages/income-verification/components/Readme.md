# Plaid Gallery Component

A react component that renders a bank gallery that triggers the Plaid interface which
allows connection with Plaid services and authentication as well as hooks and events
suitable for most interactions.

## Dependencies

In order to use this component as intended some things need to happen previously:
1) On server-side node, prepare the config needed for plaid and pass it to the front end
2) On client-side, load the plaid-initialize script
3) Create a Plaid Client and pass it to the component

As well as [Snowflake](https://github.tlcinternal.com/pages/UITools/ui-snowflake/docs/versions/1.6.x/build/index.html) to be imported for basic styles.
Classes needed from Snowflake are:
- `col-xs-6`, `col-sm-4` Used for grid positioning and grid width for bank gallery.
- `margin-top-5`, `padding-all-10`, `padding-bottom-0` To add margins/paddings for the bank search control container.
- `text-center` To center content inside the bank search control container.
- `h4` for the "Find your bank" text inside search control container.
- `visible-sm-inline` for the "Even local banks in your hometown" text to dissapear on mobile and display it on desktop only.

For step 1 you need to use both `pbs-client` and `plaid` packages.
Refer to `income-verification-server.js` for a detailed implementation.

With those packages you need to pass these data to the client using
`dynamic-module-registry`:
``` js
// some-page-server.js at the '/' route get level method
get(req, res) {
  // ...
  context.plaidConfig = JSON.stringify({
    clientName: 'LendingClub',
    product: 'auth',
    selectAccount: true,
    env: 'here you need to use plaid package data',
    key: 'as well as this key, from plaid package data'
  });
  return res.render('sope-page/some-page-template', context);
}
```

And for passing it to the client-side:
``` js
{{#defineModule "plaidConfig"}}
  {{{plaidConfig}}}
{{/defineModule}}
```

For step 2 you need to load this script and have it ready for step 3
```
https://cdn.plaid.com/link/v2/stable/link-initialize.js
```

There are several ways to do it, a recommended way to do this is to include it
inside the partial *pageScripts* that renders the scripts for the page, like this:

``` js
// some-template.hbs
{{#partial "pageScripts"}}
  <script src="https://cdn.plaid.com/link/v2/stable/link-initialize.js"></script>
  {{LCScript "runtime"}}
  {{LCScript "common"}}
  {{LCScript "some-page"}}
{{/partial}}
```

For step 3, once the script is loaded, it places a `window.Plaid` object which has
a `create()` method, use it to create a Plaid client and then pass it to the component
like this:

``` js
import 'base-layout';
import React from 'react';
import { render } from 'react-dom';
import SomeComponent from './SomeComponent';
import registry from 'dynamic-module-registry';

const plaidConfig = registry.get('plaidConfig'); // assuming this is as described on step 1
render(<SomeComponent
    plaidConfig={plaidConfig}
    plaidClient={window.Plaid}
  />, document.getElementById('someComponentContainer'));
```

Once you have everything in place, create an instance of `plaidClient`
to pass it to `PlaidGallery` component like this:

``` js
constructor(props) {
  super(props);
  this.createPlaidClient();
}

/**
 * Creates a plaid client using plaid API, external config and handlers.
 */
createPlaidClient() {
  const {
    plaidClient,
    plaidConfig
  } = this.props;

  if (plaidClient && !this.plaidInstance) {
    this.plaidInstance = plaidClient.create({
      ...plaidConfig,
      onSuccess: this.handlePlaidSuccess // handler to control plaid success event
    });
  }
}
```

And finally pass this instance to the component on the render method:

``` js
// SomeComponent.js
// ...
render() {
  handleBeforeOpen = () => { }; // your own implementation

  return (
    <div>
      <PlaidGallery
        onBeforeOpen={this.handleBeforeOpen}
        plaidInstance={this.plaidInstance}
      />
    </div>
  );
}
```

## Overview

The `PlaidGallery` component renders a bank gallery component using the `PlaidLink`
component as a button to trigger the interaction. But all event-logic should be
provided through configuration in the parent component as stated in the previous section.

If you are going to interact with Plaid UI in a different way than the `PlaidGallery`,
you can still take advantage of the `PlaidLink` component on your new component as
an encapsulated way to render buttons that trigger Plaid interactions if needed.

### Props

Prop               | Type         | Required | Default value
------------------ | ------------ | -------- | --------------------------------------
`onBeforeOpen`     | `function`   | `false`  | `null`
`plaidInstance`    | `object`     | `true`   | `null`
`institutions`     | `array`      | `false`  | Refer to the below value *

* Default value for `institutions` prop:
``` json
"institutions": [
  {
    "institution": "chase",
    "ariaLabel": "chase bank logo"
  }, {
    "institution": "wells",
    "ariaLabel": "wells fargo logo"
  }, {
    "institution": "bofa",
    "ariaLabel": "bank of america logo"
  }, {
    "institution": "pnc",
    "ariaLabel": "pnc bank logo"
  }, {
    "institution": "us",
    "ariaLabel": "us bank logo"
  }, {
    "institution": "usaa",
    "ariaLabel": "usaa logo"
  }
]
```


### Props description

- `onBeforeOpen` is a function to invoke once you have clicked on the `PlaidLink` button,
  it uses a fire and forget approach.
- `plaidInstance` is the Plaid interaction object, it provides an `open` method which `PlaidLink` is triggering to launch Plaid interactions.
- `institutions` is the bank gallery bank information to be used for rendering, those can be provided and those will be used
in a dinamically generated grid, if you plan on providing a different set of banks, logos should be provided as well in css
classnames, refer to the `PlaidGallery.less` for details on logo setup.

### State

No state is tracked or needed in this component.

## Usage

### Import / Require component

In your `.js` file:
``` js
import PlaidGallery from './PlaidGallery';
```

_Note_: Remember to import [Snowflake](https://github.tlcinternal.com/pages/UITools/ui-snowflake/docs/versions/1.6.x/build/index.html) for the basic CSS as well.

### Examples

With minimal config:
```js
// some-react-client-component.js
<PlaidGallery
  plaidInstance={this.plaidInstance}
/>
```

And with a handler to trigger on plaid button click:
```js
<PlaidGallery
  onBeforeOpen={this.handleBeforeOpen}
  plaidInstance={this.plaidInstance}
/>
```
