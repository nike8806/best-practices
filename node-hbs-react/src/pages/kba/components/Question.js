import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Question extends Component {
  static propTypes = {
    type: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    answers: PropTypes.arrayOf(PropTypes.string).isRequired,
    onAnswered: PropTypes.func,
    disabled: PropTypes.bool
  };

  static defaultProps = {
    disabled: false,
    onAnswered: null
  };

  state = {
    // eslint-disable-next-line react/destructuring-assignment, react/no-unused-state
    type: this.props.type,
    answer: null
  };

  static getDerivedStateFromProps(props, state) {
    if (state.type !== props.type) {
      return {
        type: props.type,
        answer: null
      };
    }

    return null;
  }

  /**
   * Sync state with user's selection
   *
   * @param {SyntheticEvent} event The React `SyntheticEvent`
   */
  handleAnswer = (event) => {
    this.setState({ answer: event.target.value });
  };

  /**
   * Invoque prop callback and clear state when clicking Next
   */
  handleNextClick = () => {
    const { onAnswered, type } = this.props;

    if (onAnswered) {
      const { answer } = this.state;
      onAnswered({
        type,
        answer
      });
    }
  };

  /**
   * Render method
   *
   * @return {ReactElement}
   */
  render() {
    const {
      title, type, answers, disabled
    } = this.props;

    const { answer: currentAnswer } = this.state;
    return (
      <div className="question">
        <h5 className="question__title todo-heading">{title}</h5>

        <ul className="question__answers margin-bottom-32 list-unstyled">
          {answers.map(answer => (
            <li key={`${type}-${answer}`} className="question__answer">
              <label
                className="question__answer-text"
                htmlFor={`${type}-${answer}-id`}
              >
                <input
                  id={`${type}-${answer}-id`}
                  type="radio"
                  name="answer"
                  value={answer}
                  className="question__answer-selection todo-radio"
                  checked={currentAnswer === answer}
                  onChange={this.handleAnswer}
                  disabled={disabled}
                />
                <span className="question__answer-radio todo-radio-selector" />
                {answer}
              </label>
            </li>
          ))}
        </ul>

        <button
          type="button"
          className="btn btn-todo center-block question__next"
          disabled={disabled || !currentAnswer}
          onClick={this.handleNextClick}
        >
          Next
        </button>
      </div>
    );
  }
}

export default Question;
