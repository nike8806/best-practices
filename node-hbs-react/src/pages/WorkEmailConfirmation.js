import React, { Component } from 'react';
import axios from 'lc-axios';
import { Notification, Loader } from 'components';
import { WorkEmail } from './components';

const RESULT_MESSAGES = {
  DO_NOT_WANT: {
    title: 'Thanks for your feedback, we will verify your employment with a different method.',
    subtitle: 'Please be sure to finish the rest of your to-do list.'
  },
  DO_NOT_HAVE: {
    title: 'No problem, thanks for letting us know.',
    subtitle: 'Please be sure to finish the rest of your to-do list.'
  },
  EMAIL_SENT: {
    title: 'Thank you for submitting your work email.',
    subtitle: 'Please check your email and confirm your email address.'
  }
};

const NOTIFICATION_ERROR_DEFAULT = {
  level: 'error',
  message: 'Something went wrong. Please try again later.'
};

/**
 * Work email confirmation component
 *
 * Handle all interactions to verify the work email from borrower.
 */
class WorkEmailConfirmation extends Component {
  state = {
    notification: null,
    resultMessage: null,
    // Possible values: workEmailSection. resultsSection, workEmailSection, optOutSection
    sectionToShow: 'workEmailSection',
    showLoader: false
  };

  /**
   * Callback executed on InputWorkEmail. Make the AJAX request to send the email.
   *
   * @param {String} email Email introduced by the user on InputWorkEmail
   */
  handleSubmit = (email) => {
    this.setState({ showLoader: true, notification: null });
    axios.post('/todo/work-email-confirmation/send', { email })
      .then(response => {
        this.setState({
          resultMessage: RESULT_MESSAGES.EMAIL_SENT,
          showLoader: false,
          sectionToShow: 'resultsSection'
        });
      })
      .catch(error => {
        this.setState({
          notification: NOTIFICATION_ERROR_DEFAULT,
          showLoader: false
        });
      });
  };

  /**
   * Handle the Optout -> Do not want give the email
   */
  handleOptoutDoNotWant = () => {
    this.submitOptOut('DO_NOT_WANT');
  }

  /**
   * Handle the Optout -> Do not have email
   */
  handleOptoutDoNotHave = () => {
    this.submitOptOut('DO_NOT_HAVE');
  }

  /**
   * Manage AJAX request providing the opt-out option and displays a notification based on response
   *
   * @param {string} reason Reason to send in the ajax petition
   */
  submitOptOut = (reason) => {
    const message = RESULT_MESSAGES[reason];
    this.setState({ showLoader: true, notification: null });

    axios.post('/todo/work-email-confirmation/optout', { reason })
      .then((response) => {
        this.setState({
          resultMessage: message,
          showLoader: false,
          sectionToShow: 'resultsSection'
        });
      })
      .catch((error) => {
        this.setState({
          notification: NOTIFICATION_ERROR_DEFAULT,
          showLoader: false
        });
      });
  };

  /**
  * Handler for display OptOut section
  */
  handleShowOptOut = () => {
    this.setState({
      sectionToShow: 'optOutSection'
    });
  }

  /**
  * Handler for display OptOut section
  */
  handleShowWorkEmail = () => {
    this.setState({
      sectionToShow: 'workEmailSection'
    });
  }

  /**
   * Render method
   *
   * @return {ReactElement}
   */
  render() {
    const { notification, sectionToShow, showLoader, resultMessage } = this.state;
    return (
      <div className="work-email-confirmation">
        {showLoader && (
          <Loader />
        )}

        {notification && (
          <Notification level={notification.level} message={notification.message} className="notification-box-plain" />
        )}
        <div className="todo is-open">
          <div className="todo-wrapper">
            <div className="todo-body work-email-confirmation__body">
              {sectionToShow === 'workEmailSection' && (
              <div className="work-email-confirmation__email-section">
                <h1 className="work-email-confirmation__title">
                  Quickly and easily verify your employment
                </h1>
                <p>
                  The fastest way to verify your employment is to provide your work email from LendingClub.
                  This email will only be used for verification purposes.
                </p>

                <WorkEmail onSubmit={this.handleSubmit} />

                <p>
                  <b>What happens next?</b>
                </p>

                <ol>
                  <li>We&apos;ll send you a one-time confirmation email (it might take a few seconds).</li>
                  <li>You click on the button in the email.</li>
                  <li>You&apos;re done!</li>
                </ol>

                <p>
                  <button className="button-link work-email-confirmation__btn-show-opt-out-section" onClick={this.handleShowOptOut}>
                    I&apos;d rather not provide my work email
                  </button>
                  <br/>
                </p>

                <p>
                  <button className="button-link work-email-confirmation__btn-optout-do-not-have" onClick={this.handleOptoutDoNotHave}>
                    I don&apos;t have a work email
                  </button>
                </p>
              </div>
              )}

              {sectionToShow === 'optOutSection' && (
                <div className="work-email-confirmation__optout-section ">
                  <h1 className="work-email-confirmation__title">
                    Just a heads up...
                  </h1>

                  <p>
                    If you decide not to provide us with your work email, it may take us
                    longer to verify your employment. We will use this email address
                    to confirm your employment, but never for marketing purposes.
                  </p>

                  <p>
                    <button className="btn btn--blue work-email-confirmation__btn-fast-way" onClick={this.handleShowWorkEmail}>
                      Go back to the faster way
                    </button>
                  </p>

                  <p>
                    <button className="button-link work-email-confirmation__btn-optout-do-not-want" onClick={this.handleOptoutDoNotWant}>
                      I understand, but I&apos;d prefer not to
                    </button>
                  </p>
                </div>
              )}

              { sectionToShow === 'resultsSection' && (
                <div className="work-email-confirmation__result-section text-center">
                    <h1 className="work-email-confirmation__title">
                      { resultMessage.title }
                    </h1>
                    <div data-container="success-logo" className="success-logo"></div>
                    <p>
                    { resultMessage.subtitle }
                    </p>
                </div>
              )}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default WorkEmailConfirmation;
