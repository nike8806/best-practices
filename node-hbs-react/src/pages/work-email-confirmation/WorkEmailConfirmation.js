import React, { Component } from 'react';
import axios from 'lc-axios';
import { Loader, SuccessMessage, Modal, Input } from 'components';
import rules from 'ui-validator/src/common-rules';

const WORK_EMAIL_REGEX = /^[A-Za-z0-9._%+-]+@(?!(?:gmail|hotmail|yahoo|aol|comcast|msn))\w+[-\.\w]*?\.\w{2,4}$/;
const RESULT_SUCCESS_MESSAGES = {
  DO_NOT_WANT: {
    title: 'Thanks for your feedback. We will verify your employment with a different method',
    subtitle: 'Please be sure to finish the rest of your task list.'
  },
  DO_NOT_HAVE: {
    title: 'No problem. Thanks for letting us know',
    subtitle: 'Please be sure to finish the rest of your task list.'
  }
};

/**
 * Work email confirmation component
 *
 * Handle all interactions to verify the work email from borrower.
 */
class WorkEmailConfirmation extends Component {

  state = {
    optoutResult: '',
    showModal: '', //  Posible values: OPTOUT_NOT_WANT, EMAIL_SENT, ERROR
    showLoader: false,
    workEmail: '',
    errorWorkEmail: null,
    touchedWorkEmail: false
  };

  /**
   * Callback executed on InputWorkEmail. Make the AJAX request to send the email.
   *
   */
  handleSubmit = () => {
    const {
      workEmail,
      errorWorkEmail,
      touchedWorkEmail
    } = this.state;

    if (!touchedWorkEmail) {
      this.setState({
        touchedWorkEmail: true
      });
    }

    if (!workEmail || errorWorkEmail) {
      return;
    }

    this.setState({
      showLoader: true,
      optoutResult: '',
      showModal: '',
      formErrors: null
    });
    axios.post('/todo/work-email-confirmation/send', { email: workEmail })
      .then(response => {
        this.setState({
          showModal: 'EMAIL_SENT',
          showLoader: false
        });
      })
      .catch(error => {
        this.setState({
          showModal: 'ERROR',
          showLoader: false
        });
      });
  };

  /**
   * Handle the Optout -> Do not want give the email
   */
  handleOptoutDoNotWant = () => {
    this.submitOptOut('DO_NOT_WANT');
  }

  /**
   * Handle the Optout -> Do not have email
   */
  handleOptoutDoNotHave = () => {
    this.submitOptOut('DO_NOT_HAVE');
  }

  /**
  * Handler for display OptOut Modal of NOT_WANT option
  */
  handleShowModalDoNotWant = () => {
    this.setState({
      showModal: 'OPTOUT_NOT_WANT'
    });
  }

  /**
   * Manage AJAX request providing the opt-out option and displays a showError based on response
   *
   * @param {string} reason Reason to send in the ajax petition
   */
  submitOptOut = (reason) => {
    this.setState({
      showLoader: true,
      optoutResult: '',
      showModal: ''
    });

    axios.post('/todo/work-email-confirmation/optout', { reason })
      .then((response) => {
        this.setState({
          optoutResult: reason,
          showLoader: false
        });
      })
      .catch((error) => {
        this.setState({
          showModal: 'ERROR',
          showLoader: false
        });
      });
  };

  /**
   * Function to handle the modal dismiss
   */
  handleDismissModal = () => {
    this.setState({
      showModal: ''
    });
  };
  /**
   * Check if the workEmail is valid and return an error if is not
   * @param {String} workEmailValue
   * @return {String} error
   */
  validateWorkEmail = (workEmailValue) => {
    let error = (rules.required({message: 'Work email is required.'}, workEmailValue));
    if (error) return error;

    error = (rules.email({message: 'Please specify a valid email address'}, workEmailValue));
    if (error) return error;

    error = !WORK_EMAIL_REGEX.test(workEmailValue);
    if (error) {
        return `Sorry, we can\'t accept addresses from popular email providers
        (gmail, hotmail, yahoo, etc.). If you don\'t have any other work email address,
        please select \'I don\'t have a work email.\'`;
    }
  }

  /**
   * Handle the workEmail changes
   * @param {event} e
   */
  handleInputWorkemailChange = (e) => {
    const {value: workEmail} = e.target;
    const errorWorkEmail = this.validateWorkEmail(workEmail);

    this.setState({
      errorWorkEmail: errorWorkEmail,
      workEmail: workEmail
    });
  };

  /**
   * Handle if the work email was touched
   */
  handleWorkEmailInputTouched = (e) => {
    const {value: workEmail} = e.target;
    const errorWorkEmail = this.validateWorkEmail(workEmail);
    this.setState({
      errorWorkEmail: errorWorkEmail,
      touchedWorkEmail: true
    });
  }

  componentDidMount(prevProps, prevState) {
    const {workEmail} = this.state;
    const errorWorkEmail = this.validateWorkEmail(workEmail);
    this.setState({
      errorWorkEmail: errorWorkEmail
    });
  }

  componentDidUpdate(prevProps, prevState) {
    const { showModal: prevShowModal } = prevState;
    const { showModal } = this.state;

    if (prevShowModal === '' && showModal !== prevShowModal &&
      (showModal === 'ERROR' || showModal === 'EMAIL_SENT')) {
      this.timeOut = setTimeout(() => {
          this.setState({
            showModal: ''
          });
      }, 5000);
    }
  }

  /**
   * Render method
   *
   * @return {ReactElement}
   */
  render() {
    const { showLoader, optoutResult, showModal, errorWorkEmail, touchedWorkEmail, workEmail } = this.state;
    const {
      title: resultTitle,
      subtitle: resultSubtitle
    } = RESULT_SUCCESS_MESSAGES[optoutResult] || {};

    const emailErrorMessage = (touchedWorkEmail && errorWorkEmail) ? errorWorkEmail : null;

    return (
      <div className="work-email-confirmation">
        <div className="todo-body">
          {!optoutResult && (
            <div className="work-email-confirmation__email-section">
              <h1 className="todo-heading big">
                Quickly and easily verify your employment
              </h1>
              <p className="todo-text">
                The fastest way to verify your employment is to provide your work email.
                This email will only be used for verification purposes.
              </p>

              <label
                htmlFor="work-email-confirmation-input"
                className="sr-only"
              >
                Work email
              </label>

              <Input
                id="work-email-confirmation-input"
                type="email"
                placeholder="enter your work email"
                inputClassName="work-email-confirmation__work-email-input visible-inline"
                className="work-email-confirmation__work-email-control visible-inline"
                required
                errorMessage={emailErrorMessage}
                value={workEmail}
                onBlur={this.handleWorkEmailInputTouched}
                onChange={this.handleInputWorkemailChange}
              />

              <button
                type="button"
                className="btn btn-todo center-block work-email-confirmation__btn-submit-email margin-bottom-32 margin-top-32"
                onClick={ this.handleSubmit }
              >
                Submit
              </button>

              <h2 className="todo-heading">
                What happens next?
              </h2>

              <ol className="todo-text">
                <li>
                  We&apos;ll send you a one-time confirmation email (it might take a few seconds).
                </li>
                <li>
                  You click on the button in the email.
                </li>
                <li>
                  You&apos;re done!
                </li>
              </ol>

              <div className="row">
                <div className="col-xs-12 col-sm-6">
                  <button
                    type="button"
                    className="btn-todo-link work-email-confirmation__btn-show-opt-out-section visible-block margin-bottom-32"
                    onClick={this.handleShowModalDoNotWant}
                  >
                    I&apos;d rather not provide my work email
                  </button>
                </div>
                <div className="col-xs-12 col-sm-6 text-right-sm">
                  <button
                    type="button"
                    className="btn-todo-link work-email-confirmation__btn-optout-do-not-have"
                    onClick={this.handleOptoutDoNotHave}
                  >
                    I don&apos;t have a work email
                  </button>
                </div>
              </div>

            </div>
          )}

          {showModal === 'EMAIL_SENT' && (
            <Modal
              className="todo-modal work-email-confirmation__modal-email-sent"
              dialogClassName="work-email-confirmation__modal-dialog"
              onDismiss={this.handleDismissModal}
            >
              <h2 className="todo-heading text-center padding-all-0">
                Email sent! To complete verification click the link in the email you receive.
              </h2>
            </Modal>
          )}

          {showModal === 'OPTOUT_NOT_WANT' && (
            <Modal
              className="todo-modal work-email-confirmation__modal-not-want"
              dialogClassName="work-email-confirmation__modal-dialog"
              onDismiss={this.handleDismissModal}
              title="Just a heads up..."
              titleClassName="todo-heading"
            >
              <p className="todo-text">
                If you decide not to provide us with your work email, it may take us
                longer to verify your employment. We will use this email address
                to confirm your employment, but never for marketing purposes.
              </p>

              <button
                className="btn btn-todo center-block work-email-confirmation__btn-fast-way margin-bottom-32"
                onClick={this.handleDismissModal}
              >
                Go back to the faster way
              </button>

              <button
                className="btn-todo-link center-block work-email-confirmation__btn-optout-do-not-want"
                onClick={this.handleOptoutDoNotWant}
              >
                I understand, but I&apos;d prefer not to
              </button>
            </Modal>
          )}

          {showModal === 'ERROR' && (
            <Modal
              className="todo-modal work-email-confirmation__modal-error"
              dialogClassName="work-email-confirmation__modal-dialog"
              onDismiss={this.handleDismissModal}
            >
              <h2 className="todo-heading text-center padding-all-0">
                Something went wrong. Please try again later.
              </h2>
            </Modal>
          )}

          {optoutResult && (
            <SuccessMessage
              title={resultTitle}
              subtitle={resultSubtitle}
            >
              <a
                className="btn btn-todo"
                href="/account/myAccount.action"
              >
                Return to task list
              </a>
            </SuccessMessage>
          )}
        </div>
        {showLoader && (
          <Loader theme="fixed" />
        )}
      </div>
    );
  }
}

export default WorkEmailConfirmation;
