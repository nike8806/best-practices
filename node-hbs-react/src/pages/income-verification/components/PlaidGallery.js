import React, { Component } from 'react';
import PropTypes from 'prop-types';
import PlaidLink from './PlaidLink';

class PlaidGallery extends Component {
  static propTypes = {
    onBeforeOpen: PropTypes.func,
    plaidInstance: PropTypes.shape({
      open: PropTypes.func.isRequired
    }).isRequired,
    institutions: PropTypes.arrayOf(PropTypes.shape({
      institution: PropTypes.string.isRequired,
      ariaLabel: PropTypes.string.isRequired
    }))
  }

  static defaultProps = {
    onBeforeOpen: null,
    plaidInstance: null,
    institutions: [
      {
        institution: 'chase',
        ariaLabel: 'chase bank logo'
      }, {
        institution: 'wells',
        ariaLabel: 'wells fargo logo'
      }, {
        institution: 'bofa',
        ariaLabel: 'bank of america logo'
      }, {
        institution: 'pnc',
        ariaLabel: 'pnc bank logo'
      }, {
        institution: 'us',
        ariaLabel: 'us bank logo'
      }, {
        institution: 'usaa',
        ariaLabel: 'usaa logo'
      }
    ]
  }

  render() {
    const {
      onBeforeOpen,
      plaidInstance,
      institutions
    } = this.props;

    return (
      <React.Fragment>
        <div className="row-flush">
          { institutions.map(({institution, ariaLabel}) => (
            <div className="col-xs-6 col-sm-4" key={ institution }>
              <PlaidLink
                institution={ institution }
                className={`plaid-gallery__bank-logo plaid-gallery__bank-logo--${institution}`}
                role="img"
                aria-label={ ariaLabel }
                onBeforeOpen={ onBeforeOpen }
                plaidInstance={ plaidInstance }
              />
            </div>))
          }
        </div>

        <div className="plaid-gallery__find-bank margin-top-5 text-center">
          <PlaidLink
            onBeforeOpen={ onBeforeOpen }
            plaidInstance={ plaidInstance }
          >
            <div className="padding-all-10">
              <div className="h4">Find your bank</div>
              <div className="todo-text padding-bottom-0">
                Search 15,000+ banks.
                {' '}
                <span className="visible-sm-inline">Even local banks in your hometown.</span>
              </div>
            </div>
          </PlaidLink>
        </div>
      </React.Fragment>
    );
  }
}

export default PlaidGallery;
