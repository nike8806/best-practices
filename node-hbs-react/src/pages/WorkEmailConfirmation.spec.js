import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import WorkEmailConfirmation from './WorkEmailConfirmation';
import { Notification } from 'components';
import axios from 'lc-axios';

// TODO: Put this configuration in just one place
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });


/**
 * Tested Component reference along the tests cases.
 */
describe('WorkEmail Confirmation Spec React component', () => {
  let component;
  beforeEach(() => {
    component = shallow(<WorkEmailConfirmation />);
  });

  it('WorkEmailConfirmation render correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('should handle submit making an AJAX request', () => {
    axios.post = jest.fn(() => Promise.resolve({ data: 'foo' }));
    component.instance().handleSubmit('some@email.tld');
    Promise.all([axios.post]).then(() => {
      expect(axios.post).toHaveBeenCalledWith('/todo/work-email-confirmation/send', { email: 'some@email.tld' });
      expect(component.update().find(Notification).length).toBe(0);
    });
  });
  it('should display an error notification if request fails', () => {
    axios.post = jest.fn(() => Promise.reject({ error: 'foo' }));
    component.instance().handleSubmit('some@email.tld');
    Promise.all([axios.post]).then(() => {
      expect(axios.post).toHaveBeenCalledWith('/todo/work-email-confirmation/send', { email: 'some@email.tld' });
      expect(component.update().find(Notification).length).toBe(1);
    });
  });

  it('should handle opt out when request response is ok', () => {
    axios.post = jest.fn(() => Promise.resolve({ data: 'foo' }));
    component.find('button.work-email-confirmation__btn-optout-do-not-have').first().simulate('click');

    Promise.all([axios.post]).then(() => {
      expect(axios.post).toHaveBeenCalledWith('/todo/work-email-confirmation/optout', { reason: 'DO_NOT_HAVE' });
      expect(component.update().find(Notification).length).toBe(0);
    });
  });

  it('should handle opt out when request fails', () => {
    axios.post = jest.fn(() => Promise.reject({ error: 'foo' }));

    component.find('button.work-email-confirmation__btn-optout-do-not-have').first().simulate('click');

    Promise.all([axios.post]).then(() => {
      expect(axios.post).toHaveBeenCalledWith('/todo/work-email-confirmation/optout', { reason: 'DO_NOT_HAVE' });
      expect(component.update().find(Notification).length).toBe(1);
    });
  });

  it('should toggle opt out options when click in do not want link', () => {
    component.find('button.work-email-confirmation__btn-show-opt-out-section').first().simulate('click');

    expect(component.update().state().sectionToShow).toBe('optOutSection');
  });
});
