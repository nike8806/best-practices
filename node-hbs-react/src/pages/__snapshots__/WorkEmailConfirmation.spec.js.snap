// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkEmail Confirmation Spec React component WorkEmailConfirmation render correctly 1`] = `
<div
  className="work-email-confirmation"
>
  <div
    className="todo is-open"
  >
    <div
      className="todo-wrapper"
    >
      <div
        className="todo-body work-email-confirmation__body"
      >
        <div
          className="work-email-confirmation__email-section"
        >
          <h1
            className="work-email-confirmation__title"
          >
            Quickly and easily verify your employment
          </h1>
          <p>
            The fastest way to verify your employment is to provide your work email from LendingClub. This email will only be used for verification purposes.
          </p>
          <WorkEmail
            onSubmit={[Function]}
          />
          <p>
            <b>
              What happens next?
            </b>
          </p>
          <ol>
            <li>
              We'll send you a one-time confirmation email (it might take a few seconds).
            </li>
            <li>
              You click on the button in the email.
            </li>
            <li>
              You're done!
            </li>
          </ol>
          <p>
            <button
              className="button-link work-email-confirmation__btn-show-opt-out-section"
              onClick={[Function]}
            >
              I'd rather not provide my work email
            </button>
            <br />
          </p>
          <p>
            <button
              className="button-link work-email-confirmation__btn-optout-do-not-have"
              onClick={[Function]}
            >
              I don't have a work email
            </button>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
